installer/ssh_cluster.go

```diff
@@ -0,0 +1,431 @@
+	if err != nil {
+		return nil, err
+	}
+	privateKey, err := x509.ParsePKCS1PrivateKey(pem)
+	if err != nil {
+		return nil, fmt.Errorf("x509: %s", err)
+	}
+	return ssh.NewSignerFromKey(privateKey)
+}
+
+func (p privateKeySigner) PublicKey() ssh.PublicKey {
+	return p.publicKey
+}
+
+func (p privateKeySigner) Sign(rand io.Reader, data []byte) (*ssh.Signature, error) {
+	signer, err := p.Decrypt()
+	if err != nil {
+		fmt.Println(err)
+		return nil, err
+	}
+	return signer.Sign(rand, data)
+}
+
+// sort []privateKeySigner with unencrupted keys first,
+// and encrypted keys with public keys before those without
+type decryptedFirst []privateKeySigner
+
+func (s decryptedFirst) Len() int      { return len(s) }
+func (s decryptedFirst) Swap(i, j int) { s[i], s[j] = s[j], s[i] }
+func (s decryptedFirst) Less(i, j int) bool {
+	if !s[i].Encrypted && s[j].Encrypted {
+		return true
+	}
+	if s[i].Encrypted && !s[j].Encrypted {
+		return false
+	}
+	return s[i].publicKey != nil && s[j].publicKey == nil
+}
+
+func (c *SSHCluster) findSSHKeySigners() (signers []privateKeySigner) {
+	keyDir := filepath.Join(config.Dir(), "..", ".ssh")
+	if stat, err := os.Stat(keyDir); err != nil || !stat.IsDir() {
+		return
+	}
+	walkFunc := func(path string, info os.FileInfo, err error) error {
+		if info.IsDir() {
+			return nil
+		}
+		if strings.HasSuffix(path, ".pub") {
+			return nil
+		}
+		data, err := ioutil.ReadFile(path)
+		if err != nil {
+			return nil
+		}
+		b, _ := pem.Decode(data)
+		if b == nil {
+			return nil
+		}
+		s := privateKeySigner{
+			base:      c.base,
+			path:      path,
+			pem:       b,
+			Encrypted: x509.IsEncryptedPEMBlock(b),
+		}
+		if s.Encrypted {
+			publicKeyPath := fmt.Sprintf("%s.pub", path)
+			if stat, err := os.Stat(publicKeyPath); err == nil && !stat.IsDir() {
+				if data, err := ioutil.ReadFile(path); err == nil {
+					pk, err := ssh.ParsePublicKey(data)
+					if err == nil {
+						s.publicKey = pk
+					}
+				}
+			}
+			signers = append(signers, s)
+			return nil
+		}
+		privateKey, err := x509.ParsePKCS1PrivateKey(b.Bytes)
+		if err != nil {
+			return nil
+		}
+		s.key = privateKey
+		signers = append(signers, s)
+		return nil
+	}
+	filepath.Walk(keyDir, walkFunc)
+	sort.Sort(decryptedFirst(signers))
+	return
+}
+
+func (c *SSHCluster) knownHostsPath() string {
+	return filepath.Join(config.Dir(), "..", ".ssh", "known_hosts")
+}
+
+func (c *SSHCluster) parseKnownHosts() error {
+	file, err := os.Open(c.knownHostsPath())
+	if err != nil {
+		return err
+	}
+	defer file.Close()
+	k, err := knownhosts.Unmarshal(file)
+	if err != nil {
+		return err
+	}
+	c.knownHosts = k
+	return nil
+}
+
+func (c *SSHCluster) fingerprintSSHKey(publicKey ssh.PublicKey) (string, error) {
+	md5Data := md5.Sum(publicKey.Marshal())
+	strbytes := make([]string, len(md5Data))
+	for i, b := range md5Data {
+		strbytes[i] = fmt.Sprintf("%02x", b)
+	}
+	return strings.Join(strbytes, ":"), nil
```

Comment:
This should use the new format which looks like `SHA256:mVPwvezndPv/ARoIadVY98vAC0g+P/5633yTC4d/wXE`